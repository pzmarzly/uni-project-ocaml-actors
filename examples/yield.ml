open Fatamorgana
let (let* ) = bind

module Expensive
: sig
  include Actor
  val cast : data cast
end
= struct
  type data = unit
  let data_format = []
  let default () = ()
  let cast () =
    print_endline "[expens] in Expensive.cast";
    return ()
end

module Client
: sig
  include Actor
  val work : data cast
end
= struct
  type data = Expensive.data pid
  let data_format = []

  let default () = spawn (module Expensive)
  let work pid =
    print_endline "[client] calling Expensive.cast";
    let* () = cast pid Expensive.cast in
    print_endline "[client] called Expensive.cast";
    print_endline "[client] calling Expensive.cast";
    let* () = cast pid Expensive.cast in
    print_endline "[client] called Expensive.cast";
    print_endline "[client] calling yield";
    let* () = yield in
    print_endline "[client] called yield";
    print_endline "[client] calling Expensive.cast";
    let* () = cast pid Expensive.cast in
    print_endline "[client] called Expensive.cast";
    return pid
end

let main =
  cast (spawn (module Client)) Client.work

let _ =
  let ex = Executor.create () in
  Executor.add ex main;
  Executor.run_tasks ex
