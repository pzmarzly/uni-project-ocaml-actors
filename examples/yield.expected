[client] calling Expensive.cast
[client] called Expensive.cast
[client] calling Expensive.cast
[client] called Expensive.cast
[client] calling yield
[expens] in Expensive.cast
[expens] in Expensive.cast
[client] called yield
[client] calling Expensive.cast
[client] called Expensive.cast
[expens] in Expensive.cast
